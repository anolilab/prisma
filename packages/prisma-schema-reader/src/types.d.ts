export interface Field {
    name: string;
    type: string;
    list: boolean;
    required: boolean;
    isId: boolean;
    unique: boolean;
    kind: "object" | "enum" | "scalar";
    map?: string;
    relationField?: boolean;
    documentation?: string;
    relation?: { name?: string; fields?: string[]; references?: string[] };
}

export interface Model {
    name: string;
    documentation?: string;
    map?: string;
    fields: Field[];
}

export interface Enums {
    name: string;
    fields: string[];
}
export type SchemaObject = { models: Model[]; enums: Enums[] };
