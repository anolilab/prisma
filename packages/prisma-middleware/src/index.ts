export { default as applySoftDeleteMiddleware } from './soft-delete-middleware';
