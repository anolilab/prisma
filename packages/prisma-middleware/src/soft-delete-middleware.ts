import { PrismaClient } from "@prisma/client";

import type { PrismaMiddleware } from "./types";

/* eslint no-param-reassign: "off", prefer-destructuring: "off", unicorn/consistent-destructuring: "off", no-undefined: "off" */
// eslint-disable-next-line import/exports-last
export const findMiddleware: PrismaMiddleware = async (parameters, next) => {
    if (parameters.action === "findUnique") {
        // eslint-disable-next-line no-param-reassign
        parameters.action = "findFirst";

        const { where } = parameters.args;
        // Clear out the filters and reapply them so we can change the shape of a composite key (WhereUniqueInput)
        // to fit the shape of what findFirst would expect (WhereInput)
        // We do that by taking anything that isn't an Object and just putting it right back in,
        // but if we encounter an object we have to iterate through its members to pull them out one level.
        parameters.args.where = {};

        Object.entries(where).forEach((argument) => {
            if (typeof argument[1] !== "object") {
                parameters.args.where[argument[0]] = argument[1];
            } else {
                Object.entries(argument[1] as Record<string, unknown>).forEach((subarg) => {
                    parameters.args.where[subarg[0]] = subarg[1];
                });
            }
        });

        parameters.args.where.deletedAt = null;
    }

    if (parameters.action === "findMany") {
        if (!parameters.args) {
            parameters.args = {};
        }

        if (parameters.args?.where !== undefined) {
            if (parameters.args.where.deletedAt === undefined) {
                parameters.args.where.deletedAt = null;
            }
        } else {
            parameters.args.where = { deletedAt: null };
        }
    }

    return next(parameters);
};

// eslint-disable-next-line import/exports-last
export const updateMiddleware: PrismaMiddleware = async (parameters, next) => {
    if (parameters.action === "updateMany" || parameters.action === "update") {
        if (!parameters.args) {
            parameters.args = {};
        }

        if (!parameters.args.data) {
            parameters.args.data = {};
        }

        parameters.args.data.updatedAt = new Date();
    }
    return next(parameters);
};

// eslint-disable-next-line import/exports-last
export const deleteMiddleware: PrismaMiddleware = async (parameters, next) => {
    if (parameters.action === "delete") {
        parameters.action = "update";
        parameters.args.data = { deletedAt: new Date() };
    }
    if (parameters.action === "deleteMany") {
        if (!parameters.args) {
            parameters.args = {};
        }
        parameters.action = "updateMany";
        if (parameters.args?.data !== undefined) {
            parameters.args.data.deletedAt = new Date();
        } else {
            parameters.args.data = { deletedAt: new Date() };
        }
    }

    return next(parameters);
};

const applySoftDeleteMiddleware = (prisma: PrismaClient) => {
    prisma.$use(findMiddleware);
    prisma.$use(updateMiddleware);
    prisma.$use(deleteMiddleware);
};

export default applySoftDeleteMiddleware;
