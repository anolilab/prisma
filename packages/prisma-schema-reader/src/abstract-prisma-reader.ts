import { existsSync, readFileSync } from "node:fs";

import type { Field } from "./types";

abstract class AbstractPrismaReader {
    protected path: string;

    protected data: string;

    public constructor(path: string) {
        this.path = path;

        this.checkIfSchemaExit();

        this.data = readFileSync(path, { encoding: "utf-8" });
    }

    protected get models() {
        return this.data.match(/\n(model(\s)[\S\s]*?})/g);
    }

    protected get enums() {
        return this.data.match(/\n(enum(\s)[\S\s]*?})/g);
    }

    // eslint-disable-next-line class-methods-use-this
    protected blockLines(block: string) {
        return block.split(/\n/).filter((v) => v);
    }

    protected checkIfSchemaExit() {
        if (!existsSync(this.path)) {
            // eslint-disable-next-line no-console
            console.error(`Error: "schema.prisma" file not found in "${this.path}"`);
            // eslint-disable-next-line unicorn/no-process-exit
            process.exit();
        }
    }

    // eslint-disable-next-line class-methods-use-this
    protected getType(line: string[]) {
        return line[1].replace("?", "").replace("[]", "");
    }

    protected getKind(type: string) {
        return this.data.includes(`enum ${type} `) ? "enum" : (this.data.includes(`model ${type} `) ? "object" : "scalar");
    }

    protected getClassName(lines: string[]) {
        return this.lineArray(lines[0])[1];
    }

    // eslint-disable-next-line class-methods-use-this
    protected lineArray(line: string) {
        return line
            .replace(/[\n\r]/g, "")
            .split(" ")
            .filter((v) => v);
    }

    // eslint-disable-next-line class-methods-use-this
    protected getMap(line: string): string | undefined {
        const value = line.match(/@map\((.*?)\)/);

        if (value) {
            return value[1].replace(/name/, "").replace(":", "").replace(" ", "").replace(/"/g, "");
        }

        // eslint-disable-next-line no-undefined
        return undefined;
    }

    // eslint-disable-next-line class-methods-use-this
    protected getRelation(line: string): { name?: string; fields?: string[]; references?: string[] } | undefined {
        const relationString = line.match(/@relation\((.*?)\)/);

        if (relationString) {
            const relation: Field["relation"] = {};
            const name = relationString[1].match(/"(\w+)"/);

            if (name) {
                // eslint-disable-next-line prefer-destructuring
                relation.name = name[1];
            }

            ["fields", "references"].forEach((item) => {
                // eslint-disable-next-line @rushstack/security/no-unsafe-regexp
                const pattern = new RegExp(`${item}:[\\s\\S]\\[(.*?)\\]`);
                const values = relationString[1].match(pattern);
                if (values) {
                    const asArray = values[1]
                        .replace(/ /g, "")
                        .split(",")
                        .filter((v) => v);

                    if (asArray.length > 0) {
                        relation[item as "fields" | "references"] = asArray;
                    }
                }
            });

            return relation;
        }

        // eslint-disable-next-line no-undefined
        return undefined;
    }
}

export default AbstractPrismaReader;
