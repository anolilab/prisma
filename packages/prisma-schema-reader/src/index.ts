export { default as AbstractPrismaReader } from "./abstract-prisma-reader";
export { default as SchemaToObjectTransformer } from "./schema-to-object-transformer";
export type {
    Field, Model, SchemaObject, Enums,
} from "./types";
