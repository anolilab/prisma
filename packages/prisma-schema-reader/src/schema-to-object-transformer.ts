import AbstractPrismaReader from "./abstract-prisma-reader";
import {
    Enums, Field, Model, SchemaObject,
} from "./types";

class SchemaToObjectTransformer extends AbstractPrismaReader {
    schemaObject: SchemaObject = {
        models: [],
        enums: [],
    };

    public transform(): SchemaObject {
        this.getModels();
        this.getEnums();

        return this.schemaObject;
    }

    // eslint-disable-next-line radar/cognitive-complexity
    private getModels() {
        if (this.models) {
            this.models.forEach((model) => {
                const lines = this.blockLines(model);
                const modelObject: Model = {
                    name: this.getClassName(lines),
                    fields: [],
                };

                let documentation = "";

                // eslint-disable-next-line no-plusplus
                for (let index = 1; index + 1 < lines.length; index++) {
                    const line = this.lineArray(lines[index]);

                    if (line[0].includes("//")) {
                        documentation = documentation ? `${documentation}\n${line.join(" ")}` : line.join(" ");
                    } else if (line[0].includes("@@")) {
                        modelObject.map = this.getMap(lines[index]);
                    } else {
                        const type = this.getType(line);
                        const field: Field = {
                            name: line[0],
                            type,
                            isId: line.includes("@id"),
                            unique: line.includes("@unique"),
                            list: line[1].includes("[]"),
                            required: !line[1].includes("[]") && !line[1].includes("?"),
                            kind: this.getKind(type),
                            documentation,
                            map: this.getMap(lines[index]),
                        };

                        if (field.kind === "object") {
                            field.relation = this.getRelation(lines[index]);
                        }

                        modelObject.fields.push(field);
                        documentation = "";
                    }
                }

                modelObject.documentation = documentation;

                modelObject.fields
                    .filter((item) => item.kind !== "object")
                    .forEach((item) => {
                        let relationField = false;

                        modelObject.fields
                            .filter((field) => field.kind === "object")
                            .forEach((field) => {
                                if (!relationField) {
                                    relationField = !!field.relation?.fields?.includes(item.name);
                                }
                            });

                        // eslint-disable-next-line no-param-reassign
                        item.relationField = relationField;
                    });

                this.schemaObject.models.push({ ...modelObject });
            });
        }
    }

    private getEnums() {
        if (this.enums) {
            this.enums.forEach((item) => {
                const lines = this.blockLines(item);
                const itemObject: Enums = {
                    name: this.getClassName(lines),
                    fields: [],
                };

                // eslint-disable-next-line no-plusplus
                for (let index = 1; index + 1 < lines.length; index++) {
                    const line = this.lineArray(lines[index]);

                    // eslint-disable-next-line @typescript-eslint/no-unused-expressions
                    !line[0].includes("//") && itemObject.fields.push(line[0]);
                }

                this.schemaObject.enums.push({ ...itemObject });
            });
        }
    }
}

export default SchemaToObjectTransformer;
