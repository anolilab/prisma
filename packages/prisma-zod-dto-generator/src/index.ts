import { generatorHandler } from "@prisma/generator-helper";
import path from "path";
import { Project, StructureKind, VariableDeclarationKind } from "ts-morph";

generatorHandler({
    onManifest() {
        return {
            prettyName: "Zod Schemas",
            defaultOutput: "zod",
            version: "0.2.1",
        };
    },
    onGenerate(options) {
        const project = new Project({
            skipAddingFilesFromTsConfig: true,
        });

        const outputPath = options.generator.output!.value;
        const { models } = options.dmmf.datamodel;

        const prismaClient = options.otherGenerators.find(
            (each) => each.provider.value === "prisma-client-js",
        );

        const { relationModel } = options.generator.config as unknown as Config;

        const indexSource = project.createSourceFile(
            `${outputPath}/index.ts`,
            {},
            {
                overwrite: true,
            },
        );

        models.forEach((model) => {
            indexSource.addExportDeclaration({
                moduleSpecifier: `./${model.name.toLowerCase()}`,
            });
        });
    },
});
