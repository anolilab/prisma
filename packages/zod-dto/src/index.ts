export { default as AbstractDto } from "./abstract-dto";
export { BooleanDto, zBooleanDto } from "./boolean-dto";
export type { BooleanDtoValue } from "./boolean-dto";
export { zCurrencyDto, CurrencyDto } from "./currency-dto";
export type { CurrencyDtoValue } from "./currency-dto";
export { zDatetimeDto, DatetimeDto } from "./datetime-dto";
export type { DatetimeDtoValue } from "./datetime-dto";
export { zEmailDto, EmailDto } from "./email-dto";
export type { EmailDtoValue } from "./email-dto";
export { zImageDto, ImageDto } from "./image-dto";
export type { ImageDtoValue } from "./image-dto";
export { zJsonDto, JsonDto } from "./json-dto";
export type { JsonDtoValue } from "./json-dto";
export { zNumberDto, NumberDto } from "./number-dto";
export type { NumberDtoValue } from "./number-dto";
export { zStringArrayDto, StringArrayDto } from "./string-array-dto";
export type { StringArrayDtoValue } from "./string-array-dto";
export { zStringDto, StringDto } from "./string-dto";
export type { StringDtoValue } from "./string-dto";
export { zUrlDto, UrlDto } from "./url-dto";
export type { UrlDtoValue } from "./url-dto";
