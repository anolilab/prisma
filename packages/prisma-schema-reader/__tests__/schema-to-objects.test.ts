import path from "path";

import { SchemaToObjectTransformer } from "../src";

const prismaSchemaPath = path.resolve("./packages/prisma-schema-reader/__tests__/fixtures/schema.prisma");

describe("schema-to-object", () => {
    test("it should transform prisma.schema to objects", () => {
        const transformer = new SchemaToObjectTransformer(prismaSchemaPath);

        expect(transformer.transform()).toMatchSnapshot();
    });
});
