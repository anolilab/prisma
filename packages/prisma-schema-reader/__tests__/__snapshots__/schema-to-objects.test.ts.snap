// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`schema-to-object it should transform prisma.schema to objects 1`] = `
Object {
  "enums": Array [
    Object {
      "fields": Array [
        "ADMIN",
        "USER",
      ],
      "name": "Role",
    },
  ],
  "models": Array [
    Object {
      "documentation": "",
      "fields": Array [
        Object {
          "documentation": "",
          "isId": true,
          "kind": "scalar",
          "list": false,
          "map": undefined,
          "name": "id",
          "relationField": false,
          "required": true,
          "type": "Int",
          "unique": false,
        },
        Object {
          "documentation": "",
          "isId": false,
          "kind": "scalar",
          "list": false,
          "map": undefined,
          "name": "createdAt",
          "relationField": false,
          "required": true,
          "type": "DateTime",
          "unique": false,
        },
        Object {
          "documentation": "",
          "isId": false,
          "kind": "scalar",
          "list": false,
          "map": undefined,
          "name": "email",
          "relationField": false,
          "required": true,
          "type": "String",
          "unique": true,
        },
        Object {
          "documentation": "",
          "isId": false,
          "kind": "scalar",
          "list": false,
          "map": undefined,
          "name": "name",
          "relationField": false,
          "required": false,
          "type": "String",
          "unique": false,
        },
        Object {
          "documentation": "",
          "isId": false,
          "kind": "scalar",
          "list": false,
          "map": undefined,
          "name": "password",
          "relationField": false,
          "required": true,
          "type": "String",
          "unique": false,
        },
        Object {
          "documentation": "",
          "isId": false,
          "kind": "object",
          "list": true,
          "map": undefined,
          "name": "posts",
          "relation": undefined,
          "required": false,
          "type": "Post",
          "unique": false,
        },
      ],
      "name": "User",
    },
    Object {
      "documentation": "",
      "fields": Array [
        Object {
          "documentation": "",
          "isId": true,
          "kind": "scalar",
          "list": false,
          "map": undefined,
          "name": "id",
          "relationField": false,
          "required": true,
          "type": "Int",
          "unique": false,
        },
        Object {
          "documentation": "",
          "isId": false,
          "kind": "scalar",
          "list": false,
          "map": undefined,
          "name": "published",
          "relationField": false,
          "required": true,
          "type": "Boolean",
          "unique": false,
        },
        Object {
          "documentation": "",
          "isId": false,
          "kind": "scalar",
          "list": false,
          "map": undefined,
          "name": "title",
          "relationField": false,
          "required": true,
          "type": "String",
          "unique": false,
        },
        Object {
          "documentation": "",
          "isId": false,
          "kind": "object",
          "list": false,
          "map": undefined,
          "name": "author",
          "relation": Object {
            "fields": Array [
              "authorId",
            ],
            "references": Array [
              "id",
            ],
          },
          "required": false,
          "type": "User",
          "unique": false,
        },
        Object {
          "documentation": "",
          "isId": false,
          "kind": "scalar",
          "list": false,
          "map": undefined,
          "name": "authorId",
          "relationField": true,
          "required": false,
          "type": "Int",
          "unique": false,
        },
        Object {
          "documentation": "",
          "isId": false,
          "kind": "scalar",
          "list": false,
          "map": undefined,
          "name": "createdAt",
          "relationField": false,
          "required": true,
          "type": "DateTime",
          "unique": false,
        },
        Object {
          "documentation": "",
          "isId": false,
          "kind": "scalar",
          "list": false,
          "map": undefined,
          "name": "updatedAt",
          "relationField": false,
          "required": true,
          "type": "DateTime",
          "unique": false,
        },
      ],
      "name": "Post",
    },
  ],
}
`;
